rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -1,2 +1,1730 @@
+                endpoint_name.to_string(),
+                DeviceStatus::default(),
+                timeout,
+            )
+            .await
+    }
+
+    async fn call_observe_device_update(
+        device_name: &str,
+        endpoint_name: &str,
+        timeout: Duration,
+        adr_client: Client<SessionManagedClient>,
+    ) -> Result<DeviceUpdateObservation, Error> {
+        adr_client
+            .observe_device_update_notifications(
+                device_name.to_string(),
+                endpoint_name.to_string(),
+                timeout,
+            )
+            .await
+    }
+
+    async fn call_unobserve_device_update(
+        device_name: &str,
+        endpoint_name: &str,
+        timeout: Duration,
+        adr_client: Client<SessionManagedClient>,
+    ) -> Result<(), Error> {
+        adr_client
+            .unobserve_device_update_notifications(
+                device_name.to_string(),
+                endpoint_name.to_string(),
+                timeout,
+            )
+            .await
+    }
+
+    #[tokio::test]
+    async fn test_get_asset_empty_asset_name() {
+        let adr_client = create_adr_client();
+        let result = call_get_asset(
+            "test-device",
+            "test-endpoint",
+            "",
+            Duration::from_secs(10),
+            adr_client,
+        )
+        .await;
+
+        assert!(matches!(
+            result.unwrap_err(),
+            Error(ErrorKind::AIOProtocolError(_))
+        ));
+    }
+
+    #[tokio::test]
+    async fn test_get_asset_empty_device_name() {
+        let adr_client = create_adr_client();
+        let result = call_get_asset(
+            "",
+            "test-endpoint",
+            "test-asset",
+            Duration::from_secs(10),
+            adr_client,
+        )
+        .await;
+
+        assert!(matches!(
+            result.unwrap_err(),
+            Error(ErrorKind::AIOProtocolError(_))
+        ));
+    }
+
+    #[tokio::test]
+    async fn test_get_asset_empty_endpoint_name() {
+        let adr_client = create_adr_client();
+        let result = call_get_asset(
+            "test-device",
+            "",
+            "test-asset",
+            Duration::from_secs(10),
+            adr_client,
+        )
+        .await;
+        assert!(matches!(
+            result.unwrap_err(),
+            Error(ErrorKind::AIOProtocolError(_))
+        ));
+    }
+
+    #[tokio::test]
+    async fn test_get_asset_zero_timeout() {
+        let adr_client = create_adr_client();
+        let result = call_get_asset(
+            "test-device",
+            "test-endpoint",
+            "test-asset",
+            Duration::from_secs(0),
+            adr_client,
+        )
+        .await;
+        assert!(matches!(
+            result.unwrap_err().kind(),
+            ErrorKind::InvalidRequestArgument(_)
+        ));
+    }
+
+    #[tokio::test]
+    async fn test_update_asset_status_empty_asset_name() {
```

Comment:
any way to use test-case to combine some of these tests? Seems like a lot of them are pretty similar